### DISCUSSION TITLE
How would you use a trait which uses setup and teardown?

### DISCUSSION THREAD

**Discussion by JasonTheAdams** [pestphp/pest#328] [Q&A]
I have the following in an existing TestCase class:
```php
    /**
     * @var array original config values
     */
    private $storedConfigs = [];

    /**
     * @var array Keys of configs to store and restore
     */
    protected $configsToStore = [];

    /**
     * {@inheritdoc}
     */
    protected function setUp(): void
    {
        parent::setUp();

        $this->storeConfigs();
    }

    /**
     * {@inheritdoc}
     */
    protected function tearDown(): void
    {
        $this->restoreConfigs();

        parent::tearDown();
    }

    /**
     * Store the configurations before the test is run to later restore them to their default state.
     */
    private function storeConfigs()
    {
        if (! empty($this->configsToStore)) {
            foreach ($this->configsToStore as $key) {
                $this->storedConfigs[$key] = config($key);
            }
        }
    }

    /**
     * Restore the ENV configurations stored before the test.
     */
    private function restoreConfigs()
    {
        if (! empty($this->configsToStore)) {
            foreach ($this->configsToStore as $key) {
                Config::set($key, $this->storedConfigs[$key]);
            }
        }
    }
```

Using this, in a Test class if I can use the `$configsToStore` property to temporarily swap mock values into the Config, and then restore those values once the test class is finished. I can't figure out the best way to make something reusable like this between test files in Pest.

Any recommendations?

**Comment by JasonTheAdams** [comment]
I tried using `Config::set` within the `beforeAll` hook and I'm getting "A facade root has not been set". It works in `beforeEach`. Is this expected behavior?

**Reply by owenvoke** [reply]
Yes, this is expected. `beforeAll()` is called before the class has been initialised (equivalent of `@beforeClass` in PHPUnit), whereas `beforeEach()` is called before each test method, but after the class has been initialised (equivalent of `setUp()` in PHPUnit).

The docs have more information about these functions: https://pestphp.com/docs/setup-and-teardown#setup-and-teardown

**Reply by JasonTheAdams** [reply]
Got it! Thanks for clarifying. Any thoughts on the overarching question to this discussion post? 😃
